pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/ControllerFilePathProvider.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.api.resources;
+
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import org.apache.commons.io.FileUtils;
+import org.apache.pinot.common.utils.CommonConstants;
+import org.apache.pinot.common.utils.URIUtils;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.filesystem.PinotFSFactory;
+import org.apache.pinot.spi.filesystem.PinotFS;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ControllerFilePathProvider {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ControllerFilePathProvider.class);
+  private static final String FILE_UPLOAD_TEMP_DIR = "fileUploadTemp";
+  private static final String UNTARRED_FILE_TEMP_DIR = "untarredFileTemp";
+  private static final String FILE_DOWNLOAD_TEMP_DIR = "fileDownloadTemp";
+
+  private static ControllerFilePathProvider INSTANCE;
+
+  /**
+   * NOTE: this should be called only once when starting the controller. We don't check whether INSTANCE is null because
+   * we might start multiple controllers in the same JVM for testing.
+   */
+  public static void init(ControllerConf controllerConf)
+      throws InvalidControllerConfigException {
+    INSTANCE = new ControllerFilePathProvider(controllerConf);
+  }
+
+  public static ControllerFilePathProvider getInstance() {
+    Preconditions.checkState(INSTANCE != null, "ControllerFilePathProvider has not been initialized");
+    return INSTANCE;
+  }
+
+  private final URI _dataDirURI;
+  private final File _fileUploadTempDir;
+  private final File _untarredFileTempDir;
+  private final File _fileDownloadTempDir;
+  private final String _vip;
```

Comment:
Rename to `_vipUrl`?

Reply:
Carried over from the old code, and not related to this PR, so would prefer keeping it not changed